src/inmanta/server/services/compilerservice.py

```diff
@@ -800,18 +800,17 @@ async def _process_next_compile_in_queue(
 
                         if (
                             # no compile is running for this environment
-                            compile.environment not in self._compiling_envs
+                            compile.environment not in self._env_to_compile_task
                             # the previously running compile is finished
                             or (finish_current_compile and compile.environment == environment)
                         ):
-                            self._compiling_envs.add(compile.environment)
                             # Run the compile. self._run will call back into this method when done
-                            self.add_background_task(self._run(compile))
+                            self._env_to_compile_task[compile.environment] = self.add_background_task(self._run(compile))
                             started_new_compile_run = True
                 finally:
                     if finish_current_compile and not started_new_compile_run:
                         assert environment is not None
-                        self._compiling_envs.discard(environment)
+                        del self._env_to_compile_task[environment]
```

Comment:
I'm not sure `del` is necessarily safe here. `discard` is "remove if exists".

Reply:
One scenario where I think it may not be in the dict is when we clear the env near the end of the compile. This function will be called, and `cancel_compile` will be called. Both will race for the lock. I don't think either can make assumptions about pre-existing state just because they've been called.